### PR TITLE
make webhook admission kind of work

### PR DESCRIPTION
**Pull request by deads2k**
This is done to support https://github.com/openshift/kubernetes-namespace-reservation/pull/3 .  It contains multiple fixes needed to make webhooks run at all.  In addition, it changes validation rules and admission handling until we get changes like https://github.com/kubernetes/kubernetes/issues/53826 into the API.  

This affects handling and compatibility of an alpha feature.

@dgoodwin @abhgupta be ready for upgrade pain in this area.
@bparees I turned this on in cluster-up.  Surgery was relatively minor

### CODE REVIEW

**Comment by bparees** [comment]
cluster up changes lgtm.

**Comment by dgoodwin** [comment]
Upgrade pain just in the sense that we'll need a build deployed with this change before we can get anything operational?

**Comment by deads2k** [comment]
> Upgrade pain just in the sense that we'll need a build deployed with this change before we can get anything operational?

Alpha fields are alpha.  The API will change and your webhook admission will either disappear or start failing.

File: pkg/cmd/server/origin/admission/plugin_initializer.go [hunk 1]
```diff
@@ -144,6 +145,14 @@ func NewPluginInitializer(
 		cloudConfig,
 		restMapper,
 		quotaRegistry)
+	// upstream broke this, so we can't use their mechanism.  We need to get an actual client cert and practically speaking privileged loopback will always have one
```
Comment (sttts) [inline pkg/cmd/server/origin/admission/plugin_initializer.go]: broke what?

File: vendor/k8s.io/kubernetes/plugin/pkg/admission/webhook/admission.go [hunk 2]
```diff
@@ -45,6 +45,8 @@ import (
 	"k8s.io/kubernetes/pkg/kubeapiserver/admission/configuration"
 
 	// install the clientgo admission API for use with api registry
+	"strings"
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/plugin/pkg/admission/webhook/admission.go]: order

File: pkg/cmd/server/origin/admission/plugin_initializer.go (hunk 1 above)
Comment (deads2k) [reply pkg/cmd/server/origin/admission/plugin_initializer.go]: > broke what?

They wire to the wrong client cert: https://github.com/kubernetes/kubernetes/issues/53828 .  Also, it's built such that using the "normal" constructor produces a plugin initializer that will panic on a "normal" admission chain.

Comment (sttts) [reply pkg/cmd/server/origin/admission/plugin_initializer.go]: That link helps. Next time I read your PR description.

File: test/integration/router/router_http_server.go [hunk 3]
```diff
@@ -19,6 +19,8 @@ import (
 	kapi "k8s.io/kubernetes/pkg/api"
 
 	"github.com/openshift/origin/pkg/cmd/util"
+
+	_ "github.com/openshift/origin/pkg/api/install"
```
Comment (sttts) [inline test/integration/router/router_http_server.go]: how is this related to the PR?

File: vendor/k8s.io/kubernetes/plugin/pkg/admission/webhook/admission.go [hunk 4]
```diff
@@ -230,8 +231,14 @@ func (a *GenericAdmissionWebhook) callHook(ctx context.Context, h *v1alpha1.Exte
...
 		return &ErrCallingWebhook{WebhookName: h.Name, Reason: err}
 	}
+
+	tokens := strings.SplitN(h.Name, "/", 2)
+	urlPath := ""
+	if len(tokens) == 2 {
+		urlPath = tokens[1]
+	}
 	response := &admissionv1alpha1.AdmissionReview{}
-	if err := client.Post().Context(ctx).Body(&request).Do().Into(response); err != nil {
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/plugin/pkg/admission/webhook/admission.go]: did this work upstream?

File: test/integration/router/router_http_server.go (hunk 3 above)
Comment (deads2k) [reply test/integration/router/router_http_server.go]: > how is this related to the PR?

if the APIs aren't installed, their fake apiserver fails.

**Comment by deads2k** [comment]
comments addressed

File: pkg/cmd/server/origin/admission/plugin_initializer.go [hunk 5]
```diff
@@ -144,6 +145,14 @@ func NewPluginInitializer(
...
+	kubePluginInitializer.SetClientCert(privilegedLoopbackConfig.TLSClientConfig.CertData, privilegedLoopbackConfig.TLSClientConfig.KeyData)
```
Comment (caesarxuchao) [inline pkg/cmd/server/origin/admission/plugin_initializer.go]: Where can a webhook find the CA that signs this cert?

Comment (deads2k) [reply pkg/cmd/server/origin/admission/plugin_initializer.go]: > Where can a webhook find the CA that signs this cert?

This cert is signed by the client CA for the apiservers.  Since the our webhook apiservers are run as aggregated API servers, they automatically trust that CA (though a user could wire it differently).  That means that normal cluster authentication and authorization (rbac) rules apply.
